contrib/mesos/pkg/minion/server.go

```diff
@@ -0,0 +1,269 @@
+
+	// prepare parameters
+	kmArgs := []string{server}
+	for _, arg := range *args {
+		kmArgs = append(kmArgs, arg)
+	}
+
+	// create command
+	cmd := exec.Command(ms.kmBinary, kmArgs...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		// fatal error => terminate minion
+		err = fmt.Errorf("error getting stdout of %v: %v", server, err)
+		ms.exit <- err
+		return err
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		// fatal error => terminate minion
+		err = fmt.Errorf("error getting stderr of %v: %v", server, err)
+		ms.exit <- err
+		return err
+	}
+
+	ch := make(chan struct{})
+	go func() {
+		defer func() {
+			select {
+			case <-ch:
+				log.Infof("killing %v process...", server)
+				if err = cmd.Process.Kill(); err != nil {
+					log.Errorf("failed to kill %v process: %v", server, err)
+				}
+			default:
+			}
+		}()
+
+		maxSize := ms.logMaxSize.Value()
+		if maxSize > 0 {
+			// convert to MB
+			maxSize = maxSize / 1024 / 1024
+			if maxSize == 0 {
+				log.Warning("maximal log file size is rounded to 1 MB")
+				maxSize = 1
+			}
+		}
+		writer := &lumberjack.Logger{
+			Filename:   logFileName,
+			MaxSize:    int(maxSize),
+			MaxBackups: ms.logMaxBackups,
+			MaxAge:     ms.logMaxAgeInDays,
+		}
+		defer writer.Close()
+
+		log.V(2).Infof("Starting logging for %v: max log file size %d MB, keeping %d backups, for %d days", server, maxSize, ms.logMaxBackups, ms.logMaxAgeInDays)
+
+		<-ch
+		written, err := io.Copy(writer, stderrLogs)
+		if err != nil {
+			log.Errorf("error writing data to %v: %v", logFileName, err)
+		}
+
+		log.Infof("wrote %d bytes to %v", written, logFileName)
+	}()
+
+	// if the server fails to start then we exit the executor, otherwise
+	// wait for the proxy process to end (and release resources after).
+	if err := cmd.Start(); err != nil {
+		// fatal error => terminate minion
+		ms.exit <- err
+		err = fmt.Errorf("error starting %v: %v", server, err)
+		ms.exit <- err
+		return err
+	}
+	close(ch)
+	if err := cmd.Wait(); err != nil {
+		log.Error("%v exited with error: %v", server, err)
+		err = fmt.Errorf("%v exited with error: %v", server, err)
+		return err
+	}
+
+	return nil
+}
+
+// runs the main kubelet loop, closing the kubeletFinished chan when the loop exits.
+// never returns.
+func (ms *MinionServer) Run(hks hyperkube.Interface, _ []string) error {
+	if ms.privateMountNS {
+		// only the Linux version will do anything
+		enterPrivateMountNamespace()
+	}
+
+	clientConfig, err := ms.KubeletExecutorServer.CreateAPIServerClientConfig()
+	if err != nil {
+		// required for k8sm since we need to send api.Binding information
+		// back to the apiserver
+		log.Fatalf("No API client: %v", err)
+	}
+	ms.clientConfig = clientConfig
+
+	// run subprocess until
+	if ms.runProxy {
+		go runtime.Until(ms.launchProxyServer, 5*time.Second, ms.done)
+	}
+	go runtime.Until(ms.launchExecutorServer, 5*time.Second, ms.done)
+
+	// wait until minion exit is requested
+	err = <-ms.exit
+	close(ms.exit)
```

Comment:
A write to a closed channel will either block (and then gets unblocked by the close) or fails because the channel is closed. No? Which race-condition do you have in mind?